src/interpolation/details/ArborX_InterpDetailsSymmetricPseudoInverseSVD.hpp

```diff
@@ -0,0 +1,231 @@
+      sin = Kokkos::sqrt(value_t(2)) / 2;
+      x = a + b;
+      y = a - b;
+    }
+    else
+    {
+      value_t const u = (2 * b) / (a - c);
+      value_t const v = 1 / Kokkos::sqrt(u * u + 1);
+      cos = Kokkos::sqrt((1 + v) / 2);
+      sin = Kokkos::copysign(Kokkos::sqrt((1 - v) / 2), u);
+      x = (a + c + (a - c) / v) / 2;
+      y = a + c - x;
+    }
+
+    // Now lets compute the following new values for U amd E.S
+    // M <- R'(theta)^T . S . R'(theta)
+    // U <- U . R'(theta)
+
+    // R'(theta)^T . S . R'(theta)
+    std::size_t i = 0;
+    for (; i < p; i++)
+    {
+      value_t const s_ip = s(i, p);
+      value_t const s_iq = s(i, q);
+      s(i, p) = cos * s_ip + sin * s_iq;
+      s(i, q) = -sin * s_ip + cos * s_iq;
+    }
+    s(p, p) = x;
+    i++;
+    for (; i < q; i++)
+    {
+      value_t const s_pi = s(p, i);
+      value_t const s_iq = s(i, q);
+      s(p, i) = cos * s_pi + sin * s_iq;
+      s(i, q) = -sin * s_pi + cos * s_iq;
+    }
+    s(q, q) = y;
+    i++;
+    for (; i < size; i++)
+    {
+      value_t const s_pi = s(p, i);
+      value_t const s_qi = s(q, i);
+      s(p, i) = cos * s_pi + sin * s_qi;
+      s(q, i) = -sin * s_pi + cos * s_qi;
+    }
+    s(p, q) = 0;
+
+    // U . R'(theta)
+    for (std::size_t i = 0; i < size; i++)
+    {
+      value_t const u_ip = u(i, p);
+      value_t const u_iq = u(i, q);
+      u(i, p) = cos * u_ip + sin * u_iq;
+      u(i, q) = -sin * u_ip + cos * u_iq;
+    }
+
+    norm = argmaxUpperTriangle(p, q);
+  }
+
+  // We compute the max to get a range of the invertible eigen values
+  value_t max = epsilon;
+  for (std::size_t i = 0; i < size; i++)
+    max = Kokkos::max(Kokkos::abs(s(i, i)), max);
+
+  // We inverse the diagonal of S, except if "0" is found
+  for (std::size_t i = 0; i < size; i++)
+    s(i, i) = (Kokkos::abs(s(i, i)) < max * epsilon) ? 0 : 1 / s(i, i);
+
+  // Then we fill out IO as the pseudo inverse
+  for (std::size_t i = 0; i < size; i++)
+    for (std::size_t j = 0; j < size; j++)
+    {
+      value_t tmp = 0;
+      for (std::size_t k = 0; k < size; k++)
+        tmp += s(k, k) * u(i, k) * u(j, k);
+      io(i, j) = tmp;
+    }
+}
+
+template <typename ExecutionSpace, typename InOutMatrices>
+void symmetricPseudoInverseSVD(ExecutionSpace const &space, InOutMatrices &ios)
+{
+  // InOutMatrices is a single or list of matrices (i.e 2 or 3D view)
+  static_assert(Kokkos::is_view_v<InOutMatrices>, "In-out data must be a view");
+  static_assert(InOutMatrices::rank == 3 || InOutMatrices::rank == 2,
+                "In-out view must be a matrix or a list of matrices");
+  static_assert(
+      KokkosExt::is_accessible_from<typename InOutMatrices::memory_space,
+                                    ExecutionSpace>::value,
+      "In-out view must be accessible from the execution space");
+  using view_t = Kokkos::View<typename InOutMatrices::non_const_data_type,
+                              typename InOutMatrices::memory_space>;
```

Comment:
This view is used to create the auxiliary matrices. Because `InOutMatrices` may not have a good layout, this forces the extra matrices to have the correct layout for the memory space.
And yes, I should have added that assertion as `ios` is written to.